umi_tools/umi_methods.py

```diff
@@ -1104,6 +1104,80 @@ def get_read_position(read, soft_clip_threshold):
 
 
+class get_readpairs:
+    # Used in place of a real read as the mate of reads that aren't properly paired
+    NoMate = object()
+
+    def incomplete_pair_to_singleton(self, queue_entry):
+        for i in [i for i, v in enumerate(queue_entry) if v is get_readpairs.NoMate]:
+            queue_entry[i] = None
+
+    def incomplete_pairs_as_singletons(self):
+        self.incomplete_queue_entries.clear()
+        while self.queue:
+            queue_entry = self.queue.popleft()
+            if get_readpairs.NoMate in queue_entry:
+                # Incomplete entry. Turn into singleton entry
+                read = next(r for r in queue_entry if r is not get_readpairs.NoMate)
+                U.warn("inconsistent BAM file: only one of the two mates of proper pair %s found on chromosome %s, treating as unpaired" %
+                       (read.qname, read.reference_name))
+                incomplete_pair_to_singleton(queue_entry)
+            yield queue_entry
+
+    def __call__(self, inreads):
+        self.queue = collections.deque()
+        self.incomplete_queue_entries = dict()
+        self.current_chr = None
+
+        for read in inreads:
+            read_i = int(read.is_read2)
+            read_chr = read.reference_name
+
+            # Output leftover incomplete read pairs if the chrosomome changes
+            if self.current_chr is not None and self.current_chr != read_chr:
+                for queue_entry in self.incomplete_pairs_as_singletons():
+                    yield queue_entry
+            self.current_chr = read_chr
+
+            # Queue read, either as a singleton, or as part of a read pair
+            if read.is_unmapped or read.mate_is_unmapped or not read.is_proper_pair:
+                # Read is not part of a proper pair. Set mate to None
+                queue_entry = [None, None]
+                queue_entry[read_i] = read
+                self.queue.append(queue_entry)
+            elif read.qname in self.incomplete_queue_entries:
+                # Read is part of an already queued (incomplete) read pair. Complete it.
+                queue_entry = self.incomplete_queue_entries.pop(read.qname)
+                if queue_entry[read_i] is not get_readpairs.NoMate:
+                    U.warn("inconsistent BAM file: both mates %s flagged to be read%d" %
+                           (read.qname, read_i+1))
+                    read_i = 1 - read_i
+                queue_entry[read_i] = read
+            else:
+                # Read is part of a new read pair. Create & queue incomplete pair.
+                queue_entry = [get_readpairs.NoMate, get_readpairs.NoMate]
+                queue_entry[read_i] = read
+                self.queue.append(queue_entry)
+                self.incomplete_queue_entries[read.qname] = queue_entry
+
+            # Output queued reads and readpairs up to the first incomplete pair
+            while self.queue and sum(1 for r in self.queue[0] if r is get_readpairs.NoMate) == 0:
```

Comment:
Yeah, this is definately going to become a mess if you have say to alignments for read1 followed by an alignment for read2. The two alignments of read one will be treated as a read one and a read two, and output as a pair, then read 2 will be encountered and a new pair will be created.

Reply:
Hm, yeah, multi-mapping breaks this. Again, I'm inclined to simply say "multi-mapping" and --paired are contradictory, but again I might be missing something...